ops.py

```diff
@@ -59,17 +59,39 @@ class GGMLLayer(torch.nn.Module):
     comfy_cast_weights = True
     dequant_dtype = None
     patch_dtype = None
-
-    def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs):
+    ggml_delegate_types = {None, gguf.GGMLQuantizationType.F32, gguf.GGMLQuantizationType.F16}
+
+    def ggml_check_delegate(self, *, weight=None, bias=None):

```

Comment:
@city96 what do you think of this approach?

we can delegate both the private Torch methods (scary to override) and op functions in `manual_cast` whenever there are no quantized tensors to deal with and hopefully make breaking stuff less likely.

could also potentially just add a handler for all the ops types without making non-quantized stuff break.

Reply:
Not bad, I think that should work. One small nitpick is that maybe `ggml_check_delegate` could be `is_quantized` instead with the logic flipped since that makes it a lot more apparent what this does. When reading over it I kept going "wait why is- oh nvm delegate means the regular ones" lmao. (also, better handle both args and kwargs for the input).

```py
    def is_quantized(self, *args, weight=None, bias=None, **kwargs):
        if weight is None:
            weight = self.weight
        if bias is None:
            bias = self.bias
        weight_type = getattr(weight, "tensor_type", None)
        bias_type = getattr(bias, "tensor_type", None) if bias is not None else None
        torch_compatible = {None, gguf.GGMLQuantizationType.F32, gguf.GGMLQuantizationType.F16}
        return weight_type not in torch_compatible and bias_type not in torch_compatible
```

Reply:
no problem about changing the name/flipping the logic. what are you doing with `*args` and `**kwargs` though? it only optionally gets passed specific weight/bias to use or nothing, it's not the actual wrapper - just a function that returns a boolean.

Reply:
i renamed the function to `is_ggml_quantized` and reversed the logic. i didn't bring the list of Torch compatible types into the function because there's no need to rebuild that set every single time the function is called.

the thinking behind calling it `is_ggml_quantized` rather than `is_quantized` is that we're subclassing both Torch modules and Comfy stuff and i want to avoid clobbering something that already exists. so i tried to choose names that aren't likely to conflict with any existing method/variable/etc.

of course, if you want to name it exactly something that's no problem and i can change it to whatever you want.

Reply:
Isn't it technically still like `*args`? since it just "consumes" any positional arguments but without assigning it to a value? I could be wrong (it also doesn't matter in this case because yeah, as you say it won't be used for anything else).

Also, what do you think about moving some of this stuff to dequant.py in the form of a quick check like this? Since we have to check this in other places like nodes.py anyway. Was thinking something like this:

![image](https://github.com/user-attachments/assets/f4b1a700-fac8-4980-b358-8096a649029e)

Reply:
> Isn't it technically still like *args? since it just "consumes" any positional arguments but without assigning it to a value? 

nope, it actually doesn't consume any arguments. it's just a marker that any following arguments must be passed as keyword arguments.

```plaintext
>>> def testy(a, *, x=None, y=None): print(a, x, y)
... 
>>> testy(1)
1 None None
>>> testy(1, x=2, y=3)
1 2 3
>>> testy(1, 2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: testy() takes 1 positional argument but 2 were given
```

you can also define a function like `def(blah, *, x, y): ...` and `x` and `y` _still_ have to be passed as keyword arguments, however since they have no default value they aren't optional: you must pass them or hit an exception.

***

> Also, what do you think about moving some of this stuff to dequant.py in the form of a quick check like this?

sure, that sounds reasonable. i think you're right that it probably makes more sense there than as a method.

Reply:
Huh. Learned something new today. You can tell all I ever used python for was random automation crap.

(Also, don't think we need to add the `_ggml_` into function names that often unless it's something that already has that. So for the load state dict it makes sense but maybe not as much for `is_quantized`? Though most of the naming in the repo is pretty random anyway)

Reply:
if left `ggml` in the method name (couldn't easily be eliminated) but refactored it based on your example. how about this iteration?

Reply:
Yeah the new changes look good, thank you.

The merge conflict is on me, though I think at least it fixed the damn double RAM/random OOM issue. LMK if you think it's ready to merge into the `ops_refactor` branch. I'll ask for a review from you when we merge that one to master to make sure I didn't break anything.

Reply:
i resolved the conflicts, still seems to work with Flux at least. i have to go for now, i won't get a chance to test other models until tomorrow unfortunately. not sure if you want to wait or not.

i've been testing with all but your most recent changes

Reply:
Sounds good, I'll merge it to this branch and do any changes on my end (i.e. switching away from mmap loading since it's broken on latest comfy anyway and doesn't save any time/memory).

We're not in any hurry so I'll just wait for your final OK before doing the merge with main whenever you're free.